src/nav2_tasks/include/nav2_tasks/TaskServer.hpp

```diff
@@ -0,0 +1,164 @@
+        std::bind(&TaskServer::onCommandReceived, this, std::placeholders::_1));
+
+    cancelSub_ = create_subscription<std_msgs::msg::String>(name + "_cancel",
+        std::bind(&TaskServer::onCancelReceived, this, std::placeholders::_1));
+
+    resultPub_ = this->create_publisher<ResultMsg>(name + "_result");
+    statusPub_ = this->create_publisher<StatusMsg>(name + "_status");
+
+    startWorkerThread();
+  }
+
+  virtual ~TaskServer()
+  {
+    stopWorkerThread();
+  }
+
+  virtual TaskStatus executeAsync(const typename CommandMsg::SharedPtr command) = 0;
+
+  // The user's executeAsync method can check if the client is requesting a cancel
+  bool cancelRequested()
+  {
+    return shouldCancel_;
+  }
+
+  void setCanceled()
+  {
+    shouldCancel_ = false;
+  }
+
+  void setResult(const ResultMsg & result)
+  {
+    resultMsg_ = result;
+  }
+
+protected:
+  // These messages are internal to the TaskClient implementation
+  typedef std_msgs::msg::String CancelMsg;
+  typedef std_msgs::msg::String StatusMsg;
+
+  // The pointer to our private worker thread
+  std::thread * workerThread_;
+
+  // This class has the worker thread body which calls the user's executeAsync() callback
+  void workerThread()
+  {
+    std::mutex m;
+    std::unique_lock<std::mutex> lock(m);
+
+    do {
+      cv_.wait_for(lock, std::chrono::milliseconds(10));
+
+      if (shouldExecute_) {
+        TaskStatus status = executeAsync(commandMsg_);
+
+        if (status == TaskStatus::SUCCEEDED) {
+          // If the task succeeded, publish the result first
+          resultPub_->publish(resultMsg_);
+
+          // Then send the success code
+          std_msgs::msg::String statusMsg;
+          statusMsg.data = "Success";
+          statusPub_->publish(statusMsg);
+        } else if (status == TaskStatus::FAILED) {
+          // Otherwise, just send the failure code
+          std_msgs::msg::String statusMsg;
+          statusMsg.data = "Failure";
+          statusPub_->publish(statusMsg);
+        } else if (status == TaskStatus::CANCELED) {
+          shouldCancel_ = false;
+        } else {
+          throw std::logic_error("Unexpected status return from task");
+        }
+
+        shouldExecute_ = false;
+      }
+    } while (rclcpp::ok());
+  }
+
+  // Convenience routes for starting and stopping the worker thread (used from the ctor and dtor)
+  void startWorkerThread()
+  {
+    workerThread_ = new std::thread(&TaskServer::workerThread, this);
+  }
+
+  void stopWorkerThread()
+  {
+    workerThread_->join();
+    delete workerThread_;
+    workerThread_ = nullptr;
+  }
+
+  std::condition_variable cv_;
+  std::atomic<bool> shouldCancel_;
+  std::atomic<bool> shouldExecute_;
+
+  typename CommandMsg::SharedPtr commandMsg_;
+  ResultMsg resultMsg_;
+
+  // The callbacks for our subscribers
+  void onCommandReceived(const typename CommandMsg::SharedPtr msg)
+  {
+    std::cout << "onCommandReceived\n";
+    commandMsg_ = msg;
+    shouldExecute_ = true;
```

Comment:
Design question - I'm not sure I understand why a thread is needed at all, couldn't the OnCommandReceived() method call the WorkerThread() method directly instead of setting a flag here?

Reply:
I'm working under the assumption that we don't want to use one of the ROS threads for doing potentially long-running work. Instead, we want to do the minimum and return the thread to ROS for handling other input messages. I don't know if there's a convention or guidance in the ROS community for this.

Reply:
I'll check what rclcpp::executor is doing. I was assuming it spawns a thread for every callback it makes, but I could be wrong.